csrc/expr_evaluator.cpp

```diff
@@ -197,6 +197,11 @@ const PolymorphicValue& ExpressionEvaluator::evaluate(ParallelType pt) {
   return null_;
 }
 
+const PolymorphicValue& ExpressionEvaluator::evaluate(const Val* value) const {
+  const auto it = known_values_.find(value);
+  return it != known_values_.end() ? it->second : null_;
+}
+
```

Comment:
This behaves a bit different from the non-const evaluate, since it is restricted to known values only so it is "shallow" like `getValue`. Could we instead consider making `getValue` const (which it should be) and further just exposing that as a public method? That way it is clear that `getValue` does a shallow but const lookup and `evaluate` does a deep but non-const one. Alternatively, we could implement evaluate this const version recursively just like the non-const one, but just avoid the step where we set values in `known_values_`.